src/SPC/builder/BuilderBase.php

```diff
@@ -233,15 +234,52 @@ public function proveExts(array $static_extensions, array $shared_extensions = [
      */
     abstract public function buildPHP(int $build_target = BUILD_TARGET_NONE);
 
+    /**
+     * Test PHP
+     */
+    abstract public function testPHP(int $build_target = BUILD_TARGET_NONE);
+
+    /**
+     * @throws WrongUsageException
+     * @throws RuntimeException
+     * @throws FileSystemException
+     */
     public function buildSharedExts(): void
     {
-        foreach ($this->getExts() as $ext) {
-            if (!$ext->isBuildShared()) {
-                continue;
+        $lines = file(BUILD_BIN_PATH . '/php-config');
+        $extension_dir_line = null;
+        foreach ($lines as $key => $value) {
+            if (str_starts_with($value, 'extension_dir=')) {
+                $lines[$key] = 'extension_dir="' . BUILD_MODULES_PATH . '"' . PHP_EOL;
+                $extension_dir_line = $value;
+                break;
             }
-            logger()->info('Building extension [' . $ext->getName() . '] as shared extension (' . $ext->getName() . '.so)');
-            $ext->buildShared();
         }
+        file_put_contents(BUILD_BIN_PATH . '/php-config', implode('', $lines));
+        FileSystem::createDir(BUILD_MODULES_PATH);
+        try {
+            foreach ($this->getExts() as $ext) {
+                if (!$ext->isBuildShared()) {
+                    continue;
+                }
+                if (Config::getExt($ext->getName(), 'type') === 'builtin' || Config::getExt($ext->getName(), 'build-with-php') === true) {
+                    if (file_exists(BUILD_MODULES_PATH . '/' . $ext->getName() . '.so')) {
+                        logger()->info('Shared extension [' . $ext->getName() . '] was already built by php-src/configure (' . $ext->getName() . '.so)');
+                        continue;
+                    }
+                    if (Config::getExt($ext->getName(), 'build-with-php') === true) {
+                        logger()->warning('Shared extension [' . $ext->getName() . '] did not build with php-src/configure (' . $ext->getName() . '.so)');
+                        logger()->warning('Try deleting your build and source folders and running `spc build`` again.');
+                        continue;
+                    }
+                }
+                $ext->buildShared();
+            }
+        } catch (RuntimeException $e) {
+            FileSystem::replaceFileLineContainsString(BUILD_BIN_PATH . '/php-config', 'extension_dir=', $extension_dir_line);
+            throw $e;
+        }
+        FileSystem::replaceFileLineContainsString(BUILD_BIN_PATH . '/php-config', 'extension_dir=', $extension_dir_line);
```

Comment:
I'm not so sure if this is appropriate. Maybe this paragraph can be simplified further.